arch/gpt: check for userspace ABI instead of CPU family

bluca opened the pull request:
spec change: https://github.com/uapi-group/specifications/pull/110

poettering commented on src/basic/architecture.c:
```diff
@@ -150,6 +151,7 @@ static const char *const architecture_table[_ARCHITECTURE_MAX] = {
         [ARCHITECTURE_CRIS]        = "cris",
         [ARCHITECTURE_X86_64]      = "x86-64",
         [ARCHITECTURE_X86]         = "x86",
+        [ARCHITECTURE_X32]         = "x32",
```
no, that's not an architecture. It's a calling convention. There are no CPUs that implement this "architecture". The CPUs are just regular x86 cpus.

Also, it's really dead afaik. Let's not add this now.

bluca replied:
these are not architectures in the cpu sense, they are architectures in the binary abi/package sense, and x32 is a different packaging architecture that is separate and independent from amd64 and i386, and has its own separate builds. So just like you have an amd64/i386 rootfs or usr partition, you also have an x32 partition that ships an installation made of x32 deb packages

bluca replied:
And it's not dead as a Debian architecture, it is in the same state as other non-release ports like riscv or loongarch or sparc64, from the Debian point of view they are all at the same level of support

bluca replied:
https://packages.debian.org/sid/x32/systemd/filelist

poettering replied:
we use the partition types to automatically find the partition to boot into based on efi arch or cpu or kernel arch we find. Nothing of that kind applies to x32. you cannot auto-select that. 

people should just use mark their x32 partitions as x86-64. It makes sense to then mark individual binaries with x32, but the file system as a whole doesn't really make any sense.

There are numerous abis around, including a myriad for arm. we never exposed that either.

poettering replied:
why not just map x32 to x86_64 on debian?

bluca replied:
You cannot mark an x32 partition as amd64, because then it would be picked up on amd64 which is wrong. The entire point of x32 is to have a different user space abi than amd64, so they have to be separate, they are not interchangeable

bluca replied:
And yes the separate selection logic needs some tweaks too - for example using efi arch to select user space can lead to wrong results too, ie there are cursed machines with ia32 firmware and amd64 CPUs that will boot am amd64 kernel and an x32 userspace, because why not

poettering replied:
> You cannot mark an x32 partition as amd64, because then it would be picked up on amd64 which is wrong. The entire point of x32 is to have a different user space abi than amd64, so they have to be separate, they are not interchangeable

I am not following.

There is no x32 kernel, there's no x32 uefi or anything. you can use the same kernel and boot either 64bit or x32 userspace with it, it's entirely fine. Hence, any auto-discovery of root partition is entirely happy with *either* of these, hence there's no point in having distinct partition types for them.

whether you use x86-64 abi or x32 abi is mostly a choice like "ubuntu" or "fedora". We don't have distinct partition types for that either.

bluca replied:
These GUIDs match the ABI they contain, not the CPU. Otherwise we would need to always mark an i386 partition as amd64, unless it is actually running on an i386 CPU, as you can take an amd64 CPU, boot an amd64 kernel, and boot an i386 partition. The ABI of the binaries contained in a x32 partition is x32, not amd64 nor i386.

poettering replied:
Nah, it works the other way round: if systematically the binaries in some partition cannot be run on any other arch then they need their own per-arch id.

hence: because there are plenty binaries that work on x86-64 but do not work on i386 they need a distinct arch type. but there are no real binaries that work on x86-64 but do not work on x32 systems and vice versa, hence they do not need and should not have a distinct identifier.

bluca replied:
"cannot be run" is a property of an individual system in an individual state, and I cannot possibly know what that will be when building an image beforehand. You can make the same argument for i386: I can run any of that on amd64, hence we can just alias it? Of course not. The only thing I know when I am building an image is, what is the architecture of the content. I do not know on which kernel or which cpu it will be run on.

poettering commented:
So I think the other stuff makes sense, but the x32 thing should not be added. People can just add use x86-64 for that, it's the actual CPU architecture for that.

poettering commented:
When i added architecture.h i thought about precisely this issue btw, see comment at the top:

```
/* A cleaned up architecture definition. We don't want to get lost in
 * processor features, models, generations or even ABIs. Hence we
 * focus on general family, and distinguish word width and endianness. */
```

i.e. I *on* *purpose* wanted to avoid any such ABI mess. Please do *not* sneak ABI information into this thing it really has no place here.

bluca commented:
I don't care about architecture.h, I only need a set of separate GPT GUIDs. The compiler is unhappy unless that is defined in architecture.h though.

poettering commented:
Sorry, but for GPT it makes no sense either. Why do you think you need that for x32? why not just map x32 to x86-64 gpt type?

bluca commented:
Because it's different and incompatible content. If I want amd64 I'll tag and pick amd64, if I want something else I'll pick something else.

bluca commented:
If I have an image with an x32 partition and I dissect it, I want the dissect tool to tell me it's x32, not amd64. Just like I want it to tell me it's armv7 and not arm64, even if I'm running an arm64 CPU that can execute armv7 binaries.

poettering commented:
> If I have an image with an x32 partition and I dissect it, I want the dissect tool to tell me it's x32, not amd64. Just like I want it to tell me it's armv7 and not arm64, even if I'm running an arm64 CPU that can execute armv7 binaries.

but why is that interesting? you can use the x32 image on any x86-64 host. Whether the binaries are linked one way or another inside of it doesn't really matter at all for anything, does it? Not for nspawn, not for portabled, not for sysext?

bluca commented:
> > If I have an image with an x32 partition and I dissect it, I want the dissect tool to tell me it's x32, not amd64. Just like I want it to tell me it's armv7 and not arm64, even if I'm running an arm64 CPU that can execute armv7 binaries.
> 
> but why is that interesting? you can use the x32 image on any x86-64 host. Whether the binaries are linked one way or another inside of it doesn't really matter at all for anything, does it? Not for nspawn, not for portabled, not for sysext?

It does matter, because the content is different and incompatible. I cannot run an x32 sysext on an amd64 rootfs. It's a different set of libraries, binaries, ABIs, etc.
Similarly as I cannot run an armv7 sysext on an arm64 host, and so on and so forth. I need an entire different set of glibc&friends.

Identifying the content of the image is what's this useful for. If it's A but the tools say B, then it's no longer useful.
Again, whether something can or cannot be run on a particular cpu/kernel combination is orthogonal - when I partition an image I have no way of knowing what it will be run on or with. I only know what the content is, so that's what I can tag it with. And A and B exist at the same time and independently, so I need different tags to distinguish them, and match them correctly.

bluca commented:
I noticed we have the same issue with arm vs armhf vs armel, which is even worse as those are release architectures, so updated to deal with that. Now when there are multiple userspace ABIs, the mapping from uname to architecture takes into consideration the userspace ABI too

poettering commented:
please *no*.

lets not turn the architecture thing into a fucking mess that just replicates what debian has gotten itself into.

poettering commented on src/basic/architecture.c:
```diff
@@ -18,7 +18,8 @@ Architecture uname_architecture(void) {
          * We try to distinguish CPUs, not CPU features, i.e. actual architectures that
-         * have genuinely different code. */
+         * have genuinely different code. This includes different userspace ABI, as used
+         * for LIB_ARCH_TUPLE and defined in https://wiki.debian.org/Multiarch/Tuples */
```
sorry, but we aren't debian, we aren't just doign whatever nonsense debian wants.

if you want, then introduce another enum that identifies abis, but please, don't just funnel all that historic crap into a currently very usable architecture field.

bluca replied:
Usable for what, exactly? Again, your suggestion of faking a userspace ABI for another just cannot work. Try and build an x32 systext, mark it as x86_64 and attach it to your x86_64 system.

poettering replied:
or actually don't, we actually have an ABI identifier already: LIB_ARCH_TUPLE! It's 1:1 the debian thing. Why do we need to duplicate that in the Architecture enum? I don't get it?

I *explicitly* designed Architecture to not be about ABI, I added comments around it, to make this clear. I am of the very strong opinion that this should not be changed.

bluca replied:
This is needed in the preprocessor macro that generates the GPT macros, so unfortunately it cannot be the LIB_ARCH_TUPLE

poettering commented:
> Similarly as I cannot run an armv7 sysext on an arm64 host, and so on and so forth. I need an entire different set of glibc&friends.

arm7 and arm64 do have different gpt/archicture enums.

bluca commented:
> please _no_.
> 
> lets not turn the architecture thing into a fucking mess that just replicates what debian has gotten itself into.

So are you suggesting Debian should stop using systemd? Because it's not like I can ignore architectures I don't like, no matter how nice that would be. Right now the architecture selection process is completely bogus and falls apart the moment a single cpu can support multiple userspace ABIs. Which is not a thing Debian invented, by the way.

The right thing to do is to consider _both_ the kernel/cpu family _and_ the glibc/gcc ABI we are built with, not ignoring one or the other

poettering commented:
if you really want to include ABI stuff in the dissection logic, then maybe make that a gpt partition flags thing. i.e. set a bunch of bits in the gpt part flags aside, that are an ABI enumerator. If not set it means "any ABI of the indicated architecture". Otherwise the bits indicate which ABIs this partition is for.

poettering commented:
> So are you suggesting Debian should stop using systemd? 

What changed that makes it so essential that Debian's ABIs all can have their own GPT partition types? that never was an issue before? why is it suddenly so crucial?

> Because it's not like I can ignore architectures I don't like, no matter how nice that would be. Right now the architecture selection process is completely bogus and falls apart the moment a single cpu can support multiple userspace ABIs. Which is not a thing Debian invented, by the way.

It doesn't fall apart though. Again, it should be fine to run nspawn or an portable service for x32 on an x86-64 host, no?

> The right thing to do is to consider _both_ the kernel/cpu family _and_ the glibc/gcc ABI we are built with, not ignoring one or the other

dunno, i don't think it's really that clear at all. abis are strange beasts, there's usually a much greater parallel support of abis for the same cpu than there is support for foreign cpu types.

I really don#t think we should consider abis and cpu types all the same thing, they are not.

bluca commented:
> > So are you suggesting Debian should stop using systemd?
> 
> What changed that makes it so essential that Debian's ABIs all can have their own GPT partition types? that never was an issue before? why is it suddenly so crucial?

Because I noticed it only now, when adding support for the DPS to the installer, and running short of GUIDs. But as mentioned many times, this is an issue that goes beyond the GPT types at this point, architecture matching for extensions and so on is also affected

> > Because it's not like I can ignore architectures I don't like, no matter how nice that would be. Right now the architecture selection process is completely bogus and falls apart the moment a single cpu can support multiple userspace ABIs. Which is not a thing Debian invented, by the way.
> 
> It doesn't fall apart though. Again, it should be fine to run nspawn or an portable service for x32 on an x86-64 host, no?

No, again, it is not fine at all. Try running an x32 portable with an x86 extension. Or run a x32 extension on your x86_64 host. You cannot do that, it's userspace ABI matching that ignores userspace ABI differences, and tries to pretend that the CPU type is the only thing that matters, but it's not.

> > The right thing to do is to consider _both_ the kernel/cpu family _and_ the glibc/gcc ABI we are built with, not ignoring one or the other
> 
> dunno, i don't think it's really that clear at all. abis are strange beasts, there's usually a much greater parallel support of abis for the same cpu than there is support for foreign cpu types.
> 
> I really don#t think we should consider abis and cpu types all the same thing, they are not.

This pretty much applies only to 3 architectures: x86, arm, mips, in the real world. Yes theoretically there could be more, but if they don't ship anywhere that we support then it doesn't matter. Adding new ABIs is a gigantic amount of work, and happens very very very rarely. It's not like something new pops out every month. But at the same time, when it does, we can't just ignore it. As long as some selection or matching based on userspace ABI is done, and it _IS_ done, then these details need to be handled, they cannot just be ignored. This doesn't mean considering CPU type and ABI the same, but none of this stuff is used solely for the CPU type, it's very much used to select and match userspace ABI too. It just does it using the CPU type, which again, it's just not right.

poettering commented:
anyway, if you want an enum for abis, then that's fine by me. i am just strictly against folding this into the Architecture enum. keep that two separate concepts.

poettering commented:
i.e. add: 

```c
typedef enum AbiType {
        ABI_PRIMARY,    /* primary abi for that architecture */
        ABI_X32,        /* on ARCHITECTURE_X86_64: the x32 ABI */
        …
} AbiType;
```

bluca commented:
I can add a new type, but I'm not sure what difference it would make? Unless I am missing something, every use of the current type that I can see would need to just switch over to the new type, as they are all related to the userspace ABI. Where would the current type still be of use?

bluca commented:
Ok I've added a separate enum, and switched almost everything to it. Anything that deals with images/files/etc is moved. Basically all that remains is calling qemu in vmspawn, which is interested exclusively in the CPU type, and the personality, same thing. It seems really really weird to add all this duplication and churn to avoid adding 3 items to an existing enum, but if that's what it takes to fix this, so be it

poettering commented:
uh, that's really not what i was proposing. I was suggesting to have a two-dimensional approach, i.e. use Architecture for the CPU architecture, but then have AbiType as an enum that selects additional abis that are not the regular ones.

poettering commented:
and also, i don't think abis should have their own gpt type uuids. instead, model this via gpt partition flags, as i suggested above.

poettering commented:
your current patch can't model the x32 abi properly. a 64bit x86 CPU after all can support three abis, not two, on the same kernel, and the SECONDARY_ABI doesn't model that.

poettering commented:
and the vpick stuff should not really take abis into consideration like this. there's no problem with running x32 and x86-64 nspawn containers and portable services and from the outside should not matter, that's an implementation detail.

bluca commented on src/basic/architecture.c:
```diff
@@ -176,3 +178,228 @@ static const char *const architecture_table[_ARCHITECTURE_MAX] = {
 
 DEFINE_STRING_TABLE_LOOKUP(architecture, Architecture);
+
+Abi uname_abi(void) {
```
I'll rename it to be more accurate - this takes into account the kernel architecture PLUS the ABI, which is what's needed, but yeah the name is not great - it was just a search+replace

bluca commented:
> uh, that's really not what i was proposing. I was suggesting to have a two-dimensional approach, i.e. use Architecture for the CPU architecture, but then have AbiType as an enum that selects additional abis that are not the regular ones.

I know that, but that suggestion doesn't work. This is not a two-dimensional thing, it's very much one-dimensional. A vendor /usr/ tree is for _one_ ABI, not multiple ones. The runner might decide to use it in several different settings, assuming it is a full tree including ld-linux (which is a big if, extensions exist), but that is orthogonal and independent of the filesystem

> and also, i don't think abis should have their own gpt type uuids. instead, model this via gpt partition flags, as i suggested above.

As above - flags do not work for this, it's a ugly hack to cover over the messiness of the current situation. The filesystem is for _one abi_, not one _cpu type_. The running system has a cpu type. These parameters are used together, but independently, and the filesystem is inherently tied to an abi, not a cpu, with the former being a more specific definition than the latter.

bluca commented:
> and the vpick stuff should not really take abis into consideration like this. there's no problem with running x32 and x86-64 nspawn containers and portable services on an x86-64 kernel and from the outside the abi used inside should not matter, that's an implementation detail.

Yes, it _has_ to - once again, you cannot run a vpick-ed armhf extension on top of an armel rootfs.

bluca commented:
> > and the vpick stuff should not really take abis into consideration like this. there's no problem with running x32 and x86-64 nspawn containers and portable services on an x86-64 kernel and from the outside the abi used inside should not matter, that's an implementation detail.
> 
> Yes, it _has_ to - once again, you cannot run a vpick-ed armhf extension on top of an armel rootfs.

and it's not just extensions btw - right now armel and armhf rootfs are marked with the same GUID as if they were interchangeable. They are not. You _cannot_ run an armhf rootfs on an armel CPU, as the entire point is that the armel CPU does not have a floating point unit. But the GUID is the same, so nspawn/pid1/portabled will happily take an armhf rootfs and try to run it on armel.

bluca commented:
Now extensions matching is fixed - if the root has an ABI defined, then the extensions are filtered on that.

This architecture situation is really messy, there are 3 separate and independent ways to indicate it: in the filename, used by vpick, in the partition GUID, used by the generator and by nspawn, and in the os-release/extension-release files, used by portabled/pid1. There are no cross-checks whatsoever, so you could have a DDI that has the ARM64 GPT UUID, is named foo_mips.raw, and contains ARCHITECTURE=x86 in os-release, and we just accept it without questions.

keszybz commented on src/basic/architecture.c:
```diff
@@ -176,3 +178,228 @@ static const char *const architecture_table[_ARCHITECTURE_MAX] = {
+         * We try to distinguish CPUs, not CPU features, i.e. actual architectures that
+         * have genuinely different code. This includes different userspace ABI, as used
+         * for LIB_ARCH_TUPLE and defined in https://wiki.debian.org/Multiarch/Tuples as
+         * ooposed to uname_architecture() above, which is solely and exclusively for CPU
```
typo

keszybz commented on src/basic/architecture.c:
```diff
@@ -176,3 +178,228 @@ static const char *const architecture_table[_ARCHITECTURE_MAX] = {
+         * for LIB_ARCH_TUPLE and defined in https://wiki.debian.org/Multiarch/Tuples as
+         * ooposed to uname_architecture() above, which is solely and exclusively for CPU
+         * comparisons. In most cases, especially when comparing images, you want to use
+         * this instead. */
```
"this instead" will not be clear without context. Maybe say "this instead of Architecture".

keszybz commented on src/partition/repart.c:
```diff
@@ -6981,7 +6981,7 @@ static int help(void) {
                "     --discard=BOOL       Whether to discard backing blocks for new partitions\n"
                "     --sector-size=SIZE   Set the logical sector size for the image\n"
-               "     --architecture=ARCH  Set the generic architecture for the image\n"
+               "     --architecture=ARCH  Set the generic abi for the image\n"
```
"--architecture=ABI" and "generic ABI"?

keszybz commented on src/shared/dissect-image.h:
```diff
@@ -44,7 +44,7 @@ struct DissectedPartition {
         ((DissectedPartition) {                                         \
                 .partno = -1,                                           \
-                .architecture = _ARCHITECTURE_INVALID,                  \
+                .abi = _ABI_INVALID,                  \
```
Align the backslash.

keszybz commented on src/basic/architecture.h:
```diff
@@ -49,6 +49,54 @@ typedef enum {
+        ABI_ALPHA,
+        ABI_ARC,
+        ABI_ARC_BE,
+        ABI_ARM,
+        ABI_ARMEL,
+        ABI_ARMHF,
+        ABI_ARM64,
+        ABI_ARM64_BE,
+        ABI_ARM_BE,
```
So there's a "arm-el" and "arm-be" variants, but also "arm"? What does that mean?

bluca replied:
ARMEL is not little endian, it's software floating point. So it's ARM, ARMEL, ARMHF, and big endian counterparts. Two reasons to keep the old ARM as-is: no idea whether it should represent hf o el, so best to be left alone, and also it could represent the even older userspace ABI which was used until (I think) armv5, and it's called arm-linux-gnu in terms of gcc/glibc ABI. It's from older releases and no longer used. Given this could not have been realistically used for arm <= 7 until these changes, and if it was I have no idea for what, I think it's best to leave the old GPT GUIDs there, and use new ones for the more specific and correct ABIs.

bluca replied:
see https://wiki.debian.org/Multiarch/Tuples for the full list

keszybz commented:
As to the overall change: this changes the meaning of the GPT UUIDs. Or more precisely, makes them more fine-grained. Previously, we lumped everything that could run on a given CPU type under one UUID, and now we'd have different UUIDs for different ABIs supported by that CPU type. I can see how this is useful.

> A vendor /usr/ tree does not have a CPU family, it has an ABI triplet.

True.

bluca commented:
> I am sorry, but this is really wrong. `/usr/` trees on Debian specifically, _of_ _course_ support multiple ABIs! it's why `/usr/lib/<triplet>` exists on Debian.

No, they do not. Libraries are just one subset of a vendor tree. /bin/sh has ONE abi. A root or usr or extension partition has nothing to do with cpu types, apart from the qemu case, everything else is about binary abi, and the current setup is completely broken and useless, and needs to be fixed. It was wrong to tie them to cpu types, as it makes no sense at all.

AdrianVovk commented:
> You cannot run an armhf rootfs on an armel CPU, as the entire point is that the armel CPU does not have a floating point unit

Well then armhf and armel are different architectures. They are not ABIs

I don't know the exact details here, but it seems to me like armhf is a superset of armel? It's the same architecture, but w/ hardware float support instead of software emulated float? So then, just like x86_64 is a superset of x86, armhf is a distinct architecture that is a superset of armel. So then armhf is an architecture with ARCHITECTURE_SECONDARY=armel set.

bluca commented:
> > You cannot run an armhf rootfs on an armel CPU, as the entire point is that the armel CPU does not have a floating point unit
> 
> Well then armhf and armel are different architectures. They are not ABIs
> 
> I don't know the exact details here, but it seems to me like armhf is a superset of armel? It's the same architecture, but w/ hardware float support instead of software emulated float? So then, just like x86_64 is a superset of x86, armhf is a distinct architecture that is a superset of armel. So then armhf is an architecture with ARCHITECTURE_SECONDARY=armel set.

Yes, I agree, they are just different architectures and should be treated as such. If you check the first version it was implemented with architectures, but then I was asked to avoid doing that and use a different enum.

AdrianVovk commented:
Those two specifically, yes, are different architectures and belong in the architecture enum. I'd clarify that the existing ARCHITECTURE_ARM uses hardware FPU (aka, armhf), and then introduce a new ARCHITECTURE_ARM_SOFTFLOAT for armle.

But other things that you call ABIs actually are ABIs and aren't architectures. x32 is a different calling convention but it's still x86_64 machine code. That's not a different architecture

AdrianVovk commented:
Let's work through this a bit more systematically.

Let's first define things clearly. Architecture is about distinguishing what will run on a CPU and what won't. You can't run x86_64 code on an x86 CPU, so x86_64 and x86 are two different architectures. Sometimes architectures are related to each-other via backwards compatibility (i.e. x86 code will run on an x86_64 machine). But even still: any given CPU has exactly one architecture (which may be a superset of some other, older architecture)

ABI is an entirely different concept: It defines the way that binaries are linked together. It's a modifier on an architecture: binaries of all the different ABIs of an architecture can run on that CPU, but binaries of different ABIs cannot be linked together.

An image with a given architecture could have multiple ABIs available in it. As you point out, _yes_: executables will be linked according to just one ABI, and the image as a whole probably has "one overall ABI" that it uses. But nothing is linking against executables! And so their presence or the ABI they're linked against has absolutely no influence over which ABIs are available in that image. ABI availability is entirely dependent on the libraries that are in the image.

I'll give a more concrete example. You could absolutely build an OS image that has both the default x86_64 ABI and the x32 ABI available in /usr/lib. All the same libraries are available, and only differ in the ABI they were linked with. Then you can make half the executables in the system link against the default ABI, and the other half of the executables link against the x32 ABI. You'll still have a functional system: just half of the executables will use one set of libraries (the ones linked with the default x86_64 ABI) and the other half will use another set of libraries (the ones linked with the x32 ABI).

So, in summary:
- Architecture and ABI are distinct concepts
	- Architecture = what will run on the CPU
	- ABI = what binaries can be linked together
- Any given image will have exactly one architecture
- Any given image of a given architecture will have and use one or more ABIs

---

Next, what does this mean practically for us?

First: representing ABIs. Since images can have multiple ABIs inside of them, it cannot be represented in GPT partition type GUIDs. As Lennart suggested, I'd set aside a range of GPT flags to represent ABI. Maybe 5 flags or so? They'd have to be interpreted depending on the architecutre the image declares in the GPT partition type (so, for example, bit 2 on an x86_64 image might mean x32, but bit 2 on an arm64 image might mean some different ARM ABI). We don't want to use up all our flags. That way, you can toggle on multiple ABIs. Bit 1 always means "default ABI", and having no bits set means that you don't care about ABI and to skip checking whether ABIs are compatible.  

Second: how do we use this ABI data? Well, it depends on what we're trying to do. Let's walk through some scenarios, and see if it makes sense to restrict by ABI or not.

- Nspawn containers: Nope! The containers come with their own entire OS image, and so don't care about the ABI of the host
- Portable services: Nope, for the same reason
- Sysexts: yes, ABI matters. A sysext can ship executables that need to be linked/loaded against host libraries. So if a sysext declares that it's an x86_64 image using the x32 ABI, then that sysext can only be loaded onto a host that has x32 libraries available for the systext to link against
- Picking rootfs at boot: nope, ABI doesn't really matter. Images of different ABIs will all boot and run on the system. The kernel is ABI-agnostic. initrds can be of a different ABI from the rootfs and still boot the system just fine. However, I see no harm in having an optional enforcement sanity-check: the initrd can verify that it's of the same ABI as the rootfs it's picking. Or it can just prefer to pick a rootfs of the same ABI as itself, but continue to boot one of a different ABI if necessary. IDK - that's up to distros that deal with this. But strictly speaking, ABI doesn't matter here.

So ultimately it sounds like mechanisms like vpick need to be extended to be ABI aware, but only optionally. The call site of mechanisms like vpick will know whether or not they care about the ABI of the images they're picking up. Other places where architectures are used will have similar situations.

bluca commented:
> But other things that you call ABIs actually are ABIs and aren't architectures. x32 is a different calling convention but it's still x86_64 machine code. That's not a different architecture

I'm really not sure it's a distinction that matters. It's a different compiler target with a different kernel with a different loader with a different glibc. It can change depending on how one chooses to change the definition, but again not sure it's useful to do so. What matters for this specific case is: can I build an and ship an image like that, and will it be different from other images so that I cannot mix and match them, eg base from one and extension from another? The answer is yes for both groups, arm vs armel vs armhf vs arm64 and x86 vs x32 vs x86_64

AdrianVovk commented:
except you absolutely can mix different ABIs together. Just because Debian chooses not to doesn't mean it's not possible.

Arch Linux has `-lib32` variants of lots of packages. I can install glibc and I can also install, into the same system, glibc-lib32. Then I can install a 32-bit program that will link against glibc-lib32. That means that I now have a system that, simultaneously, is using two different ABIs: the x86_64 default one, and the x86 default one. Same x86_64 kernel.

So yes, you absolutely can mix and match.

AdrianVovk commented:
> I'm really not sure it's a distinction that matters

It matters because ignoring it is too restrictive. Just because the host system chooses to link its binaries against the x32 ABI doesn't mean that my nspawn containers can't choose to use the default x86_64 ABI. With what you're proposing, you can't use an x86_64 ABI nspawn container on an x32 system because the ABIs don't match up, but that's silly because the container is by definition not linking against anything on the host system.

bluca commented:
> Since images can have multiple ABIs inside of them, it cannot be represented in GPT partition type GUIDs. As Lennart suggested, I'd set aside a range of GPT flags to represent ABI.

Not only it can be represented as a different GUID, but it must. This notion of adding _yet another_ separate set of identifiers solves no problem, serves no purpose, and only creates complications. It is a backward incompatible change, requires changing all producers and all consumers, and applies to ALL architectures - because you can create a Debian image containing binaries or libraries from ALL 21 architectures at once! So what do we do, go back to a single generic "Linux" GPT GUID and 21x2 (because endianess) bit flags? If the reason is for mixing multiple ABIs into a single image, then this can be the only answer.

But it's absurd of course, because the point of this specification and implementation was _never_ to identify any possible combination of binaries in an image. The point of this specification and implementation is to let the code in the initrd auto discover which partition to boot from the disk where the ESP was found. And you cannot boot multiple ABIs. Your /sbin/init and your /bin/sh will have _one_ ABI. And that's what the GPT GUID is for, to identify which one that is. That's what everyone who uses this, uses it for, including our tools and services. So that the auto-gpt generator in the initrd can find the right partition, and the initrd code can mount it, and then switch root and execute /sbin/init. That's the ABI that matters, for the purpose of tagging a partition. You can add any other binaries or libraries you like using multi-arch or any other mechanism, that's fine, but irrelevant here. It never mattered before, and I don't see why it should start to matter now.

YHNdnzj commented:
> The point of this specification and implementation is to let the code in the initrd auto discover which partition to boot from the disk where the ESP was found. And you cannot boot multiple ABIs. Your /sbin/init and your /bin/sh will have one ABI.

Sorry, but I fail to see why we'd prohibit switching into e.g. a x86_64 late userspace from x86 initrd. UEFI has been doing this for long, if you consider those devices with x86 firmware but x86_64 CPU and OS. And I also disagree that /sbin/init shall represent the whole system. It's totally fine to have a x86_64 /sbin/init bringing up the rest of x86 userspace.

Let's not go to the other extreme, please. The ABI does matter in some scenarios, yet claiming it's the _only identifier_ and _can only be one_ won't get us anywhere either.

bluca commented:
> > The point of this specification and implementation is to let the code in the initrd auto discover which partition to boot from the disk where the ESP was found. And you cannot boot multiple ABIs. Your /sbin/init and your /bin/sh will have one ABI.
> 
> Sorry, but I fail to see why we'd prohibit switching into e.g. a x86_64 late userspace from x86 initrd. UEFI has been doing this for long, if you consider those devices with x86 firmware but x86_64 CPU and OS. And I also disagree that /sbin/init shall represent the whole system. It's totally fine to have a x86_64 /sbin/init bringing up the rest of x86 userspace.
> 
> Let's not go to the other extreme, please. The ABI does matter in some scenarios, yet claiming it's the _only identifier_ and _can only be one_ won't get us anywhere either.

It's not about prohibiting? It's about finding the right option. The SECONDARY (and now TERTIARY and QUATERNARY) are there for that, so nothing is prohibited. It is how it works already today, this changes nothing in that regard, if anything there are more fallbacks, so what you want is enhanced by this PR.

AdrianVovk commented:
I fail to see the importance of the ABI of /bin/sh. It's an exectuable.

AdrianVovk commented:
> It's not about prohibiting? It's about finding the right option. The SECONDARY (and now TERTIARY and QUATERNARY) are there for that, so nothing is prohibited. It is how it works already today, this changes nothing in that regard, if anything there are more fallbacks, so what you want is enhanced by this PR.

SECONDARY is an entirely static concept and does not help here.

Let's say I have an x32 initrd. I want to boot an x86_64 rootfs. How would I do that? You cannot assert that SECONDARY=x86_64 will work, because it's entirely possible to boot an x32 rootfs that has no support for the default x86_64 ABI. At that point, SECONDARY would claim that your x32 rootfs can run x86_64 binaries but it cannot.

If you don't include SECONDARY=x86_64, then you're making the opposite assertion that can be equally wrong. My x32 rootfs can have x86_64 libraries available. But since x32 is a seemingly unrelated architecture, there's no way to represent this situation

AdrianVovk commented:
Honestly, SECONDARY gets it kinda wrong even in the current implementation. I can have an x86_64 rootfs, and someone gives me an i686 sysext. systemd will say that "well x86_64 CPU is compatible with x86 code..." (correct assertion) "and so I can assume that this code can run on this system" (INCORRECT assertion! The ABIs don't match)

AdrianVovk commented:
> It is a backward incompatible change

No? All the old stuff would just not declare what ABIs it supports and systemd would default to "all of them", like it has been doing so far

keszybz commented:
> Let's first define things clearly. Architecture is about distinguishing what will run on a CPU and what won't. You can't run x86_64 code on an x86 CPU, so x86_64 and x86 are two different architectures. Sometimes architectures are related to each-other via backwards compatibility (i.e. x86 code will run on an x86_64 machine). But even still: any given CPU has exactly one architecture (which may be a superset of some other, older architecture)

That comment as a whole is very good. It clearly summarizes the ideas that have been raised so far in the discussion.

There is an additional nuance though: different CPUs of an "architecture" can differ by supported instructions, for example x86-64-v1, x86-64-v2, x86-64-v3, x86-64-v4 baselines, and all the other amd64 variants. So if we defined "architecture" as distinct classes of CPUs on which a specific set of instructions will execute, we'd have thousands of "architectures". In the common use of the word, we group those into "amd64"/"x86_64", but strictly speaking, code compiled for x86-64-v4 is equally incompatible with both an arm CPU and an early amd64 Athlon.

I'm bringing this up, because, if our goal is to mark partitions for "what will run on a given CPU", neither the CPU "architecture", nor Luca's proposal that lists ABIs, is actually enough. This is not theoretical. For example, recent RHEL versions switched to x86-64-v3 baseline: any image built with RHEL9 supports x86-64-v2, any RHEL10 will not run on that, it requires x86-64-v3. I expect that we'll see more of this, because AMD64 "architecture" is becoming old and the pressure to build for different baselines will only increase. A similar situation is brewing for RISCV: new baseline are being debated and soon we'll have variants of common distributions that only support newer CPUs. I don't think we should introduce a significant change that describes the major divisions in architectures that were relevant 10 years ago, but not so much now.

> Second: how do we use this ABI data? Well, it depends on what we're trying to do. Let's walk through some scenarios, and see if it makes sense to restrict by ABI or not.
> 
>  * Nspawn containers: Nope! The containers come with their own entire OS image, and so don't care about the ABI of the host
> 
>  * Portable services: Nope, for the same reason
> 
>  * Sysexts: yes, ABI matters. A sysext can ship executables that need to be linked/loaded against host libraries. So if a sysext declares that it's an x86_64 image using the x32 ABI, then that sysext can only be loaded onto a host that has x32 libraries available for the systext to link against
> 
>  * Picking rootfs at boot: nope, ABI doesn't really matter.  [...]

This means that the second patch does the wrong thing.

The only case where the ABI matters is systexts, but it's also not enough. For a sysext to be match for the host, we generally require it to be very close, e.g. having specific libraries and programs installed in specific versions. So the match for sysexts needs to be more finegrained anyway.

We have `os-release` *inside* of the filesystem inside the image. Maybe we should allow that to be more accessible without opening of the image and parsing of the file system. Or maybe we should define some different way to expose additional metadata. Either way, I don't think the format needs to be extensible and offer more than just a handful of flags.

---

OK, I was supportive of this PR, but the follow-up discussion moved me into camp negative. This would be a lot of churn, and would complicate the meaning of the flags that are currently used, but it still wouldn't actually solve the problems.  Lennart, Mike, and Adrian all seem opposed, and there have been no positive comments apart from mine, which I withdraw. I'll close this. If we are to merge something, it'll need to look significantly different, and it'll be better to have a fresh discussion.

poettering commented:
> We have `os-release` _inside_ of the filesystem inside the image. Maybe we should allow that to be more accessible without opening of the image and parsing of the file system. 

We need to authenticate it somehow. hence we stick this in the fs, so that it is covered by dm-verity and its signature.

I don't think it's strictly essential to authenticate the ABI though (after all, not using the right one should not open vulnerabilities mostly, but just just result in the image not booting), hence my suggestion to just add it to the gpt flags field.

> Or maybe we should define some different way to expose additional metadata. Either way, I don't think the format needs to be extensible and offer more than just a handful of flags.

I am curious how ELF encodes these archs and ABIs btw. if it covers all these things, then i guess we could just directly include the ELF ABI id in a subset of the GPT partition flags.

keszybz commented:
>  GPT partition flags.

This is not nearly enough. It's just a few dozen bits of information at most, and there is no sane way to squish all the information that we'd want to expose into this space.

> We need to authenticate os-release somehow

I don't think so. We could expose os-release or something different as an unauthenticated field to allow selection. We would only need to check authentication if we actually decide to use the partition.

bluca commented:
> > GPT partition flags.
> 
> This is not nearly enough. It's just a few dozen bits of information at most, and there is no sane way to squish all the information that we'd want to expose into this space.

Yes, which is why this just needs to be another GUID. Nearly infinite space, and already exists, no need to make up something new.

> > We need to authenticate os-release somehow
> 
> I don't think so. We could expose os-release or something different as an unauthenticated field to allow selection. We would only need to check authentication if we actually decide to use the partition.

Having to open every single image to figure out whether it can be used would not only be incredibly slow and cumbersome, but I don't think it would work, because you need to run after udev to do that, but the generator runs before.

AdrianVovk commented:
> This is not nearly enough. It's just a few dozen bits of information at most

I mean there aren't _that_ many ABIs, are there? As far as I can tell, for any given architecture there are only 2 or 3 ABIs relevant to Linux.

I think x86 has the most ABIs: x86_64's standard ABI, the x32 ABI, and x86's standard ABI (which it inherits because x86 is ARCH_SECONDARY for x86_64). So that's 3 ABIs available.

I don't see a problem with having only a couple dozen bits of information available. The bits are defined per GPT partition type UUID. So, the bits can depend on the overall architecture of the image. We can allocate 4 bits for ABIs. That lets us have 4 different ABIs per architecture, which is plenty. We can also reserve (but not allocate for anything) the next 4 bits, just in case.

> There is an additional nuance though: different CPUs of an "architecture" can differ by supported instructions, for example x86-64-v1, x86-64-v2, x86-64-v3, x86-64-v4 baselines, and all the other amd64 variants. So if we defined "architecture" as distinct classes of CPUs on which a specific set of instructions will execute, we'd have thousands of "architectures". In the common use of the word, we group those into "amd64"/"x86_64", but strictly speaking, code compiled for x86-64-v4 is equally incompatible with both an arm CPU and an early amd64 Athlon.

I mean yeah there's a line to draw somewhere. I'd say that the "common architecture baselines" are a good line to draw. I see no reason that we shouldn't redefine `x86_64` as the v1 baseline, then have distinct v2, v3, and v4 architectures, all with unique partition type UUIDs. I suppose the only issue might be detecting the architecture level in UEFI mode?

Ditto with the new RISCV baseline. It should be defined as a distinct architecture, maybe with ARCH_SECONDARY set to the original RISCV architecture (if they are backwards compatible)

We don't have to get so granular with all the architectures, only the relevant ones. For example, for x86 we don't need to dig into i386 vs i486 vs i586, since realistically it doesn't matter for modern distros anymore

bluca commented:
> The only case where the ABI matters is systexts, but it's also not enough.

Not really. Once again, the purpose of all this is for the code in the initrd to be able to identify which rootfs to jump into, without manually defining /etc/fstab locally. That's literally the reason this exists. This means that what matters is, which of these partitions has a /sbin/init I can run?
As things stand currently, you boot on an arm machine, and this will happily pick an armhf rootfs when running on armel, and everything explodes.

This is broken. The GUIDs need to match reality, and reality is that there need to be more GUIDs than CPU families, because the CPU family doesn't matter.

keszybz commented:
> Once again, the purpose of all this is for the code in the initrd to be able to identify which rootfs to jump into

OK, so let's go back to this. The discussion has clarified the following: the CPU instruction set used by the code matters for this, but the ABI / calling convention does not. Thus there's no disagreement about adding additional *architectures* to this list. The first version of this pull request added m86k, sparc64, sh4, *and* x32. The first three are architectures, we all agree. The latter is an ABI, and a rootfs with x32 code will work everywhere where a root with amd64 code will work. (Or in other words, the x32 rootfs uses the same amd64 instruction set.)  So x32 should not be in the architecture table.

Can you submit the part of the first patch (https://github.com/systemd/systemd/commit/916783ba214d6cc6635cf8d12742dd33bb3fe71b) without x32? We can merge that part and make *some* progress.

Then there's the case of arm/armel/armhf. Formally, those are not different architectures, but variants of the same architecture. I can see why we could want to enumerate them as separate architectures, but I also see Lennart's point that this all very old stuff and we don't want to touch this mess. At this point it's mostly Debian that still has some support. Is Debian really going to be releasing images with multiple rootfses for ARM CPU variants? This seems doubtful, in which case the existing classification is enough.

poettering commented:
> This is not nearly enough. It's just a few dozen bits of information at most, and there is no sane way to squish all the information that we'd want to expose into this space.

Can you elaborate on what kind of information you actually want to squish in there? My understanding is that most distros just focus on a few select ABIs, and those are all we need to define values for.

poettering commented on src/basic/architecture.h:
```diff
@@ -49,6 +49,54 @@ typedef enum {
+        ABI_ARM64,
+        ABI_ARM64_BE,
+        ABI_ARM_BE,
+        ABI_ARMEL_BE,
```
i thought ARMEL stands for "ARM endian little", but what is a BE version of that even supposed to be

poettering commented:
Judging from this commit and the comments here I think we hould be able to get away with:

```
typedef enum Abi {
        /* x86-64 sub abis/architectures for ARCHITECTURE_X86_64 */
        ABI_X86_64_DEFAULT = 0,
        ABI_X86_64_X32,
        ABI_X86_64_V2,
        ABI_X86_64_V3,
        ABI_X86_64_V4,
   
        /* arm subarchs for ARCHITECTURE_ARM */
        ABI_ARM_DEFAULT = 0,
        ABI_ARM_HF,
} Abi;
```

Which if i count correctly needs 3 bits to encode. If we give it 5 we should have enough room for a long time, and the gpt spec easily grants as 5 bit for this purpose

keszybz commented:
> > This is not nearly enough. It's just a few dozen bits of information at most, and there is no sane way to squish all the information that we'd want to expose into this space.
> 
> Can you elaborate on what kind of information you actually want to squish in there? My understanding is that most distros just focus on a few select ABIs, and those are all we need to define values for.

I expect that different people will want to put slightly different things. Distros do focus on a few select ABIs, that's true, and actually usually just a small handful. *If* we agree that this is only about broadly defined baselines, e.g. for AMD64, we define x86-64-v1, x86-64-v2, x86-64-v3, x86-64-v4, x86S, and we curate the list ourselves, then 5 bits i.e. 32 values should be enough for a while. But I'm not sure if 5 bits is enough to describe currently known variants of Aarch64. I'd like to see some input from people who use non-AMD64 architectures and have more practical experience with this.

AdrianVovk commented:
> ```
>typedef enum Abi {
>        /* x86-64 sub abis/architectures for ARCHITECTURE_X86_64 */
>        ABI_X86_64_DEFAULT = 0,
>        ABI_X86_64_X32,
>        ABI_X86_64_V2,
>        ABI_X86_64_V3,
>        ABI_X86_64_V4,
>   
>        /* arm subarchs for ARCHITECTURE_ARM */
>        ABI_ARM_DEFAULT = 0,
>        ABI_ARM_HF,
>} Abi
>```

>  If we agree that this is only about broadly defined baselines, e.g. for AMD64, we define x86-64-v1, x86-64-v2, x86-64-v3, x86-64-v4, x86S, and we curate the list ourselves, then 5 bits i.e. 32 values should be enough for a while. But I'm not sure if 5 bits is enough to describe currently known variants of Aarch64

The x86_64 architecutre levels are **not** ABIs! They are architectures. I cannot execute x86_64_v4 code on an x86_64_v2 CPU. Similarly, hardware float vs software float with Arm. You can't execute hardware-float code on a CPU that lacks a floating-point unit.

Since these are architectures, they should be represented with the architecture UUID types. Not via the ABI bits.

> then 5 bits i.e. 32 values should be enough for a while

With ABIs, it needs to be a bit field. So 5 bits = 5 ABIs, not 32. This is because an image can have multiple ABIs in it
